cirq-aqt/cirq_aqt/aqt_sampler.py

```diff
@@ -40,16 +41,68 @@ class AQTSampler(cirq.Sampler):
     runs a single circuit or an entire sweep remotely
     """
 
-    def __init__(self, remote_host: str, access_token: str):
+    def __init__(self, workspace: str, resource: str, access_token: str, remote_host: str = "https://arnica.aqt.eu/api/v1/"):
         """Inits AQTSampler.
 
         Args:
-            remote_host: Address of the remote device.
-            access_token: Access token for the remote api.
+            workspace: the ID of the workspace you have access to.
+            resource: the ID of the resource to run the circuit on.
+            access_token: Access token for the AQT API.
+            remote_host: Address of the AQT API.
         """
+        self.workspace = workspace
+        self.resource = resource
         self.remote_host = remote_host
         self.access_token = access_token
 
+    @staticmethod
+    def fetch_resources(access_token: str, remote_host: str = "https://arnica.aqt.eu/api/v1/") -> None:
+        """Lists the workspaces and resources that are accessible with access_token.
+
+        Prints a table to STDOUT containing the workspaces and resources that the
+        passed access_token gives access to. The IDs in this table can be used to
+        submit jobs using the run and run_sweep methods.
+
+        The printed table contains four columns:
+            - WORKSPACE ID: the ID of the workspace. Use this value to submit circuits.
+            - RESOURCE NAME: the human-readable name of the resource.
+            - RESOURCE ID: the ID of the resource. Use this value to submit circuits.
+            - D/S: whether the resource is a (D)evice or (S)imulator.
+
+        Args:
+            access_token: Access token for the AQT API.
+            remote_host: Address of the AQT API. Defaults to "https://arnica.aqt.eu/api/v1/".
+
+        Raises:
+            RuntimeError: If there was an unexpected response from the server.
+        """
+        headers = {"Authorization": f"Bearer {access_token}", "SDK": "cirq"}
+        url = urljoin(remote_host if remote_host[-1] == "/" else remote_host + "/", "workspaces")
+
+        response = get(url, headers=headers)
+        if response.status_code != 200:
+            raise RuntimeError('Got unexpected return data from server: \n' + str(response.json()))
+        
+        workspaces = cast(list, response.json())
+        col_widths = [19, 21, 20, 3]
+
+        for workspace in workspaces:
+            col_widths[0] = max(col_widths[0], len(workspace['id']))
+            for resource in workspace['resources']:
+                col_widths[1] = max(col_widths[1], len(resource['name']))
+                col_widths[2] = max(col_widths[2], len(resource['id']))
+
+        print("+-" + col_widths[0]*"-"+ "-+-" + col_widths[1]*"-" + "-+-" + col_widths[2]*"-"  + "-+-" + col_widths[3]*"-" + "-+")
```

Comment:
We considered this, but we currently need the user to manually choose the workspace and resource they want to use. By allowing these to be chosen programmatically somehow, it could introduce errors that might lead to confusion. There is also limited information currently provided by the API with which one would programmatically choose - it is literally just a list of acceptable workspace/resource IDs.

This is intended purely as a helper function to allow users to look up the ID of resources that they already have access to, in case they forgot to write them down or something.

Reply:
I still don't think this is good design.  Having a Workspace class that contains names and ids of resources would be a natural way to encapsulate this information returned from the API, provide a good place to document what is returned, be more testable, and be more testable.  You could then have a print() (or just put it in __str__) function on this class to print the information as it is here.  It also would not take too much work to move this into a class.

Reply:
I certainly agree it's not good design. Have adapted slightly to expose a fetch as well as a print static method. 

Used TypedDict instead of classes to stay consistent with what was already there. Note sure in hindsight that that was the best idea.